far/common/segment.hpp

```diff
@@ -0,0 +1,113 @@
+﻿#ifndef SEGMENT_HPP_DC38F7F7_0E1B_486E_8368_2ACE6415D234
+#define SEGMENT_HPP_DC38F7F7_0E1B_486E_8368_2ACE6415D234
+#pragma once
+
+/*
+segment.hpp
+
+*/
+/*
+Copyright © 2025 Far Group
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+1. Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+2. Redistributions in binary form must reproduce the above copyright
+   notice, this list of conditions and the following disclaimer in the
+   documentation and/or other materials provided with the distribution.
+3. The name of the authors may not be used to endorse or promote products
+   derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+//----------------------------------------------------------------------------
+
+template<typename T>
+struct segment_sentinel_t
+{
+	T m_end;
+
+	explicit operator T() const noexcept { return m_end; }
+};
+
+template<typename T>
+struct segment_length_t
+{
+	T m_length;
+
+	explicit operator T() const noexcept { return m_length; }
+};
+
+template<typename T>
+class segment_t
+{
+public:
+	segment_t() noexcept = default;
+
+	segment_t(T const Start, segment_sentinel_t<T> const End) noexcept
+		: segment_t{ Start, static_cast<T>(End) }
+	{}
+
+	segment_t(T const Start, segment_length_t<T> const Length) noexcept
+		: segment_t{ Start, Start + static_cast<T>(Length) }
+	{}
```

Comment:
I think it's a good idea to force the user to be explicit when creating `segment`s. It's a bit mouthful but eliminates guesswork when you see or use the `(int, int)` constructor.